tools/probers/bazelcache/bazelcache.go

```diff
@@ -0,0 +1,276 @@
+	d, buf, err := digestGenerator.RandomDigestBuf(*blobSize)
+	if err != nil {
+		return status.UnknownErrorf("failed to generate random data: %s", err)
+	}
+
+	// Upload via ByteStream
+	log.Infof("[ByteStream] Uploading %d bytes (digest: %s/%d)", len(buf), d.GetHash(), d.GetSizeBytes())
+	resourceName := digest.NewCASResourceName(d, *instanceName, repb.DigestFunction_SHA256)
+	reader := bytes.NewReader(buf)
+
+	_, bytesUploaded, err := cachetools.UploadFromReader(ctx, bsClient, resourceName, reader)
+	if err != nil {
+		return status.UnknownErrorf("failed to upload blob: %s", err)
+	}
+	log.Infof("[ByteStream] Upload successful: %d bytes", bytesUploaded)
+
+	// Download via ByteStream
+	log.Infof("[ByteStream] Downloading blob...")
+	var downloadBuf bytes.Buffer
+	err = cachetools.GetBlob(ctx, bsClient, resourceName, &downloadBuf)
+	if err != nil {
+		return status.UnknownErrorf("failed to download blob: %s", err)
+	}
+
+	// Verify data integrity
+	if !bytes.Equal(buf, downloadBuf.Bytes()) {
+		return status.DataLossError("downloaded data doesn't match uploaded data")
+	}
+	log.Infof("[ByteStream] Download successful and verified: %d bytes", downloadBuf.Len())
+
+	return nil
+}
+
+func probeActionCache(ctx context.Context, acClient repb.ActionCacheClient) error {
+	log.Infof("[ActionCache] Starting probe...")
+
+	// Generate a random action digest
+	digestGenerator := digest.RandomGenerator(time.Now().UnixNano())
+	actionDigest, _, err := digestGenerator.RandomDigestBuf(100)
+	if err != nil {
+		return status.UnknownErrorf("failed to generate random digest: %s", err)
+	}
+
+	// Create a simple ActionResult
+	actionResult := &repb.ActionResult{
+		ExitCode: 0,
+		StdoutRaw: []byte("test stdout"),
+		StderrRaw: []byte("test stderr"),
+	}
+
+	// Store ActionResult
+	log.Infof("[ActionCache] Storing ActionResult (action digest: %s/%d)", actionDigest.GetHash(), actionDigest.GetSizeBytes())
+	updateReq := &repb.UpdateActionResultRequest{
+		InstanceName:   *instanceName,
+		ActionDigest:   actionDigest,
+		ActionResult:   actionResult,
+		DigestFunction: repb.DigestFunction_SHA256,
+	}
+	_, err = acClient.UpdateActionResult(ctx, updateReq)
+	if err != nil {
+		return status.UnknownErrorf("failed to update action result: %s", err)
+	}
+	log.Infof("[ActionCache] ActionResult stored successfully")
+
+	// Retrieve ActionResult
+	log.Infof("[ActionCache] Retrieving ActionResult...")
+	getReq := &repb.GetActionResultRequest{
+		ActionDigest:   actionDigest,
+		InstanceName:   *instanceName,
+		DigestFunction: repb.DigestFunction_SHA256,
+	}
+	retrievedResult, err := acClient.GetActionResult(ctx, getReq)
+	if err != nil {
+		return status.UnknownErrorf("failed to get action result: %s", err)
+	}
+
+	// Verify the retrieved result
+	if retrievedResult.GetExitCode() != actionResult.GetExitCode() {
+		return status.DataLossError("exit code mismatch")
+	}
+	if !bytes.Equal(retrievedResult.GetStdoutRaw(), actionResult.GetStdoutRaw()) {
```

Comment:
heh, i guess we're defaulting to inlining this? wonder how much data that's wasting